notification-extension/ITBNotificationServiceExtension.swift

```diff
@@ -16,129 +16,138 @@ import UserNotifications
-        var contentHandlerCalled = false
-        contentHandlerCalled = loadAttachment(itblDictionary: itblDictionary)
-        
-        if !contentHandlerCalled {
-            DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
-                if let bestAttemptContent = self.bestAttemptContent {
-                    contentHandler(bestAttemptContent)
-                }
-            }
-        }
+        retrieveAttachment(itblDictionary: itblDictionary, onFailureHandler: {
+            self.lastAttemptCalled()
+        })
     }
     
     @objc override open func serviceExtensionTimeWillExpire() {
         // Called just before the extension will be terminated by the system.
         // Use this as an opportunity to deliver your "best attempt" at modified content, otherwise the original push payload will be used.
-        if let contentHandler = contentHandler, let bestAttemptContent = bestAttemptContent {
-            contentHandler(bestAttemptContent)
-        }
+        
+        lastAttemptCalled()
     }
     
-    private func loadAttachment(itblDictionary: [AnyHashable: Any]) -> Bool {
-        guard let attachmentUrlString = itblDictionary[JsonKey.Payload.attachmentUrl] as? String else { return false }
-        guard let url = URL(string: attachmentUrlString) else { return false }
-        
-        let downloadTask = URLSession.shared.downloadTask(with: url, completionHandler: { [weak self] location, response, error in
-            guard let strongSelf = self else { return }
-            
-            if error == nil, let response = response, let responseUrl = response.url, let location = location {
-                let tempDirectoryUrl = FileManager.default.temporaryDirectory
-                var attachmentIdString = UUID().uuidString + responseUrl.lastPathComponent
-                if let suggestedFilename = response.suggestedFilename {
-                    attachmentIdString = UUID().uuidString + suggestedFilename
-                }
-                
-                var attachment: UNNotificationAttachment?
-                let tempFileUrl = tempDirectoryUrl.appendingPathComponent(attachmentIdString)
-                do {
-                    try FileManager.default.moveItem(at: location, to: tempFileUrl)
-                    attachment = try UNNotificationAttachment(identifier: attachmentIdString, url: tempFileUrl, options: nil)
-                } catch { /* TODO: FileManager or attachment error */ }
-                
-                if let attachment = attachment, let bestAttemptContent = strongSelf.bestAttemptContent, let contentHandler = strongSelf.contentHandler {
-                    bestAttemptContent.attachments.append(attachment)
-                    contentHandler(bestAttemptContent)
-                }
-            } else { /* TODO: handle download error */ }
-        })
+    // MARK: - Private
+    
+    private func retrieveAttachment(itblDictionary: [AnyHashable: Any], onFailureHandler: @escaping () -> Void) {
```

Comment:
I like the name `failureHandler` instead of `onFailureHandler` if it is a closure. I am not sure that we need to pass along a closure? Can't we just call the `lastAttemptCalled` method? Also, maybe also rename `lastAttemptCalled` to `onFailure` or `onCompletion`.

Reply:
These are good suggestions; I wasn't sure if the handler needed to have that flexibility of passing along a closure, so I kept it in. I think you're right that there's no need to do that.